File: 189.rotate-array.py
#
# @lc app=leetcode id=189 lang=python3
#
# [189] Rotate Array
#
from typing import List
# @lc code=start


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        width = len(nums)
        k %= width
        nums.reverse()
        for i in range(k//2):
            nums[i], nums[k-1-i] = nums[k-1-i], nums[i]
        for i in range((width-k)//2):
            nums[k+i], nums[width-1-i] = nums[width-1-i], nums[k+i]
        # nums[:k] = nums[:k][::-1]
        # nums[k:] = nums[k:][::-1]

# @lc code=end

File: 189.rotate-array.cpp
/*
 * @lc app=leetcode id=189 lang=cpp
 *
 * [189] Rotate Array
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    void rotate(vector<int> &nums, int k)
    {
        int len = nums.size();
        k %= len;
        if (k == 0)
        {
            return;
        }
        int term = len / k - 1;
        for (int i = 0; i != term; ++i)
        {
            for (int j = 0; j != k; ++j)
            {
                auto pos = len - 1 - k * i - j;
                swap(nums[pos], nums[pos - k]);
            }
        }
        auto left = len % k, right = k;
        auto pos = 0;
        while (left != 0 && right != 0)
        {
            for (int j = 0; j != right; ++j)
            {
                swap(nums[pos + left + j], nums[pos + j]);
            }
            pos += right;
            int val = right % left;
            left -= val;
            right = val;
        }

        return;
    }
};
// @lc code=end
